lib/vector-common/src/internal_event/cached_event.rs

```diff
@@ -0,0 +1,97 @@
+use std::{
+    collections::BTreeMap,
+    sync::{Arc, RwLock},
+};
+
+use super::{InternalEventHandle, RegisterInternalEvent};
+
+/// Metrics (eg. `component_sent_event_bytes_total`) may need to emit tags based on
+/// values contained within the events. These tags can't be determined in advance.
+///
+/// Metrics need to be registered and the handle needs to be held onto in order to
+/// prevent them from expiring and being dropped (this would result in the counter
+/// resetting to zero).
+/// `CachedEvent` is used to maintain a store of these registered metrics. When a
+/// new event is emitted for a previously unseen set of tags an event is registered
+/// and stored in the cache.
+pub struct RegisteredEventCache<Event: RegisterTaggedInternalEvent> {
+    cache: Arc<
+        RwLock<
+            BTreeMap<
+                <Event as RegisterTaggedInternalEvent>::Tags,
+                <Event as RegisterInternalEvent>::Handle,
+            >,
+        >,
+    >,
+}
+
+/// This trait must be implemented by events that emit dynamic tags. `register` must
+/// be implemented to register an event based on the set of tags passed.
+pub trait RegisterTaggedInternalEvent: RegisterInternalEvent {
+    /// The type that will contain the data necessary to extract the tags
+    /// that will be used when registering the event.
+    type Tags;
+
+    fn register(tags: &Self::Tags) -> <Self as RegisterInternalEvent>::Handle;
+}
+
+/// Deriving `Clone` for `Cached` doesn't work since the `Event` type is not clone,
+/// we can happily implement our own `clone` however since we are just cloning
+/// the `Arc`.
+/// Worth noting that this is a cheap clone since the cache itself is stored behind
+/// an `Arc`.
+impl<Event: RegisterTaggedInternalEvent> Clone for RegisteredEventCache<Event> {
+    fn clone(&self) -> Self {
+        Self {
+            cache: Arc::clone(&self.cache),
+        }
+    }
+}
```

Comment:
Using `Derivative` here might help auto-generate the clone.

Reply:
Perfect!